### PR TITLE
feat(data-table): add mobile-optimized variant

### LINKED ISSUE #2938: [Table] Mobile version
### Summary

What might a table look like on mobile (small breakpoint) screens? Currently, it's as-is but with the expectation the user will scroll. But that requires a lot of work for the user to complete tasks involving comparing columns. Instead, my proposal transposes columns per row to be more scannable (and scrolling vertically only):

image

I've implemented it using angular, but I understand we'd have to splice the css and some layout from there into a separate PR to the components package. This is just a preview for now:
https://github.com/werdnanoslen/carbon-components-angular/tree/feat(table)-add-mobile-variant

### Justification

Improved usability and accessibility for mobile users.

### Desired UX and success metrics

- A user can complete tasks involving multiple columns more quickly.
- A user can access information from columns past the first few more quickly.

### Specific timeline issues / requests

No rush, just wanted to discuss this.

### Available extra resources

I made it in Angular, and I maybe could learn enough React to do that too.

### PR DESCRIPTION
**Pull request by werdnanoslen**
Closes #2938 and #3006

What might a table look like on mobile (small breakpoint) screens? Currently, it's as-is but with the expectation the user will scroll. But that requires a lot of work for the user to complete tasks involving comparing columns. Instead, my proposal transposes columns per row to be more scannable (and scrolling vertically only).

A couple things I wasn't able to do very well that I could use some help with:
- When hovering over the expanded row, there is an animation change that doesn't go exactly right. I'm not very good with animations, so I just did what I could. Check it out to see what I mean and please let me know how to fix it?
- I tried making this work without that handlebars helper (`isFirstDataTd`), but maybe there's a better way...

#### Changelog

**New**

- Files for small table variant
- Handlebars helper

#### Testing / Reviewing

Open http://localhost:3000/component/data-table--small or https://deploy-preview-2992--the-carbon-components.netlify.com/component/data-table--small.html and resize the element/window to Carbon's small breakpoint.

### CODE REVIEW

**Comment by jnm2377** [comment]
Hi @werdnanoslen 👋 Thanks so much for submitting this PR. I think we may need some design input first to determine what the mobile view of data table should look like. I'm not sure if our designers have thought this through, but perhaps a designer you work with could also contribute some design specs for this. My concern is that there seems to be a `thead` still even though the `th`'s text is now included with the `td`. Additionally I have some questions regarding design patterns for this, like should this behavior be only for default tables, is it ok for expandable tables? Should it be used with pagination? Etc. 

cc: @designertyler @aagonzales

**Comment by shixiedesign** [comment]
This is an interesting approach. Currently mobile data table is expected to scroll horizontally. We don't have bandwidth for component enhancement this project cycle but will look into this in the next one. Making an issue to track:

File: packages/components/tools/templates.js [hunk 1]
```diff
@@ -84,6 +84,22 @@ const loadContents = glob =>
   getContents(glob).then(contents => {
     contents.forEach((content, templateName) => {
       Handlebars.registerPartial(templateName, content);
+      Handlebars.registerHelper('isFirstDataTd', function(columns, index) {
+        var firstDataTdIndex = 0;
+        for (var i = 0; i < columns.length; ++i) {
+          if (
+            columns[i].section |
```
Comment (asudoh) [inline packages/components/tools/templates.js]: Did you mean `||`?

File: packages/components/tools/templates.js [hunk 2]
```diff
@@ -84,6 +84,22 @@ const loadContents = glob =>
...
+            columns[i].checkbox |
+            columns[i].additionalIcons |
+            columns[i].menu
+          ) {
+            ++firstDataTdIndex;
```
Comment (asudoh) [inline packages/components/tools/templates.js]: Did you want to bail out the loop here given you seems to be looking for the first column meeting the condition?

File: packages/components/src/components/data-table/data-table.hbs [hunk 3]
```diff
@@ -223,8 +242,14 @@
...
             <!-- all other td's -->
             {{else}}
-              <td {{#if column.numerical}} class="{{@root.prefix}}--data-table--numerical-cell" {{/if}}>
+              <td class="{{#if column.numerical}}{{@root.prefix}}--data-table--numerical-cell{{/if}}{{#if @root.isSmall}} td-small{{/if}}">
                 <!-- truncated new markup -->
+                {{#if @root.isSmall}}
+                  <label class="bx--label" for="{{column.title}}-{{@../index}}">
+                    {{column.title}}
+                  </label>
+                  <div id="{{column.title}}-{{@../index}}">
```
Comment (asudoh) [inline packages/components/src/components/data-table/data-table.hbs]: `id` based on column title tend to cause ID duplication. A unique ID, probably auto-generated, should be used (e.g. as a prefix)

**Review by asudoh — COMMENTED** [review]
Thanks @werdnanoslen for jumping in!

File: packages/components/src/components/data-table/_data-table-small.scss [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+@import '../../globals/scss/helper-mixins';
+@import '../../globals/scss/vendor/@carbon/elements/scss/import-once/import-once';
+@import '../../globals/scss/vars';
+
+/// Data table small style
+/// @access private
+/// @group data-table
+@mixin data-table-small {
+  .#{$prefix}--data-table--small--min-width {
+    min-width: initial;
```
Comment (emyarod) [inline packages/components/src/components/data-table/_data-table-small.scss:18]: I believe `initial` value for CSS rules is unsupported by IE

File: packages/components/src/components/data-table/_data-table-small.scss [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  }
+  .#{$prefix}--data-table--small {
+    tbody .td-small {
+      border: none !important;
```
Comment (emyarod) [inline packages/components/src/components/data-table/_data-table-small.scss]: can we avoid `!important`? we should be able to increase the specificity in other ways

File: packages/components/tools/templates.js (hunk 2 above)
Comment (werdnanoslen) [reply packages/components/tools/templates.js]: No, I'm waiting for all the above conditions to be false. Currently, the first `<td>` that includes normal table data is not given its own class or other identifier, it's just what's left in the final `else` block. It's also constructed in a sort of predictable way, where there aren't menus or checkboxes in between those normal `<td>`s. So the loop should bail as soon as I find the first column that doesn't meet _any_ of those conditions, hence `else { break; }`.

**Comment by werdnanoslen** [comment]
Thanks everyone, I fixed most of things posted here in this commit.

@jnm2377 the `thead` is preserved for the select-all and expand-all functions only. Otherwise, like for a simple table, the thead could be omitted. Let me know if that's the way to go and I'll update the code. And yes, I built this using the more complex table to show how it could work with all those functions, not just default tables. And I'm open to design crit as well :)

File: packages/components/tools/templates.js (hunk 1 above)
Comment (werdnanoslen) [reply packages/components/tools/templates.js]: fixed in https://github.com/carbon-design-system/carbon/pull/2992/commits/5fe1847a7841cd3c474cb413a8ed1b69a7cc618a

File: packages/components/src/components/data-table/data-table.hbs (hunk 3 above)
Comment (werdnanoslen) [reply packages/components/src/components/data-table/data-table.hbs]: fixed in https://github.com/carbon-design-system/carbon/pull/2992/commits/5fe1847a7841cd3c474cb413a8ed1b69a7cc618a

File: packages/components/src/components/data-table/_data-table-small.scss (hunk 4 above)
Comment (werdnanoslen) [reply packages/components/src/components/data-table/_data-table-small.scss:18]: fixed in https://github.com/carbon-design-system/carbon/pull/2992/commits/5fe1847a7841cd3c474cb413a8ed1b69a7cc618a

File: packages/components/src/components/data-table/_data-table-small.scss (hunk 5 above)
Comment (werdnanoslen) [reply packages/components/src/components/data-table/_data-table-small.scss]: fixed in https://github.com/carbon-design-system/carbon/pull/2992/commits/5fe1847a7841cd3c474cb413a8ed1b69a7cc618a

**Comment by jnm2377** [comment]
Thanks for the quick responses and updates @werdnanoslen. Just waiting on some design input.

**Comment by aagonzales** [comment]
I think this is ok to contribute in as a variant, atm we don't want to make it the default responsive behavior for all tables because I'm unsure if it works for all/most scenarios. 

I am seeing it break though when I actually select something. It's returning to the column format after selection.

File: packages/components/tools/templates.js (hunk 2 above)
Comment (asudoh) [reply packages/components/tools/templates.js]: Did you mean `return index === columns.findIndex(column => !column.checkbox && !column.additionalIcons && !column.menu)`?

**Comment by werdnanoslen** [comment]
@aagonzales I'm unable to replicate that behavior on my local or netlify instances. Can you record it?

File: packages/components/tools/templates.js [hunk 6]
```diff
@@ -85,7 +85,10 @@ const loadContents = glob =>
     contents.forEach((content, templateName) => {
       Handlebars.registerPartial(templateName, content);
       var uids = [];
-      Handlebars.registerHelper('uid', function() {
+      Handlebars.registerHelper('uid', function(repeatLastUid) {
+        if (repeatLastUid === 'repeat') {
```
Comment (asudoh) [inline packages/components/tools/templates.js]: I'd suggest not to do "give me the last one" approach, given tends to be brittle considering distant id-ref source/target. To make my suggestion clearer a bit, I'm suggesting that we set a unique ID in `.config.js`.

Comment (werdnanoslen) [reply packages/components/tools/templates.js]: Oh I see now. So would that be a unique id for the table, and then form the `for` here with that id + row + column?

Comment (asudoh) [reply packages/components/tools/templates.js]: Right. Please note that row has an ID and column has unique name (distinct from displayed title).

Comment (werdnanoslen) [reply packages/components/tools/templates.js]: Got it, but not all rows had an ID, so I used its index as the row identifier: https://github.com/carbon-design-system/carbon/pull/2992/commits/1a220bbc1279ef57ec6a0e5e90b3cf8235e4e52d#diff-99c667b13a41db61508c656d9474e9c7R251

File: packages/components/tools/templates.js (hunk 2 above)
Comment (werdnanoslen) [reply packages/components/tools/templates.js]: yes, that's definitely a more succinct way to put it :)
fixed in https://github.com/carbon-design-system/carbon/pull/2992/commits/104c1b9cafbffe7cdd4b12f20e3b5b4aadce40c7

**Comment by jnm2377** [comment]
> I am seeing it break though when I actually select something. It's returning to the column format after selection.

@aagonzales that's actually an existing bug in the data table demo, not just this PR. All the select rows use the same configuration for the columns, so it means that they don't actually have unique id's for the checkboxes. so if you select a row on any other variation other than the default select, it'll jump to the default select table because that's the first example with those id's.

On another note though, I am seeing some style bugs on the mobile example. It looks like when you hover on the first row, there's a transparent border, and when not hovering, the `th`s bottom borders aren't aligned. Here are some pictures of what I mean:

Screen Shot 2019-06-24 at 12 11 45 PM
Screen Shot 2019-06-24 at 12 11 36 PM

> I think this is ok to contribute in as a variant, atm we don't want to make it the default responsive behavior for all tables because I'm unsure if it works for all/most scenarios.

To add to @aagonzales 's comment. It still seems weird to me to have an empty `thead`. I know that this was kept in order to still have the expand all and select all controls, but it seems to me like those particular variations of mobile table would need some more design work. Imo, this would probably be a good solution for just default tables at the moment. Thoughts, @aagonzales?

**Comment by werdnanoslen** [comment]
> To add to @aagonzales 's comment. It still seems weird to me to have an empty `thead`. I know that this was kept in order to still have the expand all and select all controls, but it seems to me like those particular variations of mobile table would need some more design work.

Fair, it might be a small usability problem in how users might expect that row to work (e.g. they might also ask themselves if content is supposed to be loading there, but not). But I think it's a more significant _cosmetic_ issue wrt less screen real estate on mobile. 

For that, perhaps we could remove the `thead` entirely and put the expand-all and select-all functions in a button icon instead?

**Comment by jnm2377** [comment]
> For that, perhaps we could remove the `thead` entirely and put the expand-all and select-all functions in a button icon instead?

Could definitely be a possible solution, in the toolbar perhaps. It would require that users include a toolbar for those variations. Wonder what design has to say about it.

**Comment by joshblack** [comment]
cc @aagonzales

**Comment by aagonzales** [comment]
I'm wary of making this the default behavior until we understand how it will impact all data-tables. I would like to add it as a modifier or variation that users can opt into. 

When does this the mobile version get triggered? Is it at a certain screen size? Is there a column minimum where this variation wouldn't be necessary?

**Comment by werdnanoslen** [comment]
@aagonzales my intention is to trigger it at the small breakpoint, where it's unlikely a user will see more than two columns (i.e. they would likely never see a full table at this width), for example:
image

~If it's feasible, I'd support some sort of check whether there are only one or two columns to keep it at the default (non-mobile) format. I think if there's just one column that should definitely happen, actually, as the mobile version with one column is no better than the default.~

**EDIT**
Actually, I take that back, because if the table's mobile format is triggered by the number of columns, and if the user can show/hide columns, then it'd be very messy looking to see it jump between mobile and default formats. I also take it back because this isn't a simple CSS transition, it's something that must be programmed in and should be treated more like a technical option that is up to the designer/programmer to build for better or worse and should be trusted to make the right decision.

**Review by aagonzales — APPROVED** [review]
Design is approved as a variant of data table.

**Comment by jnm2377** [comment]
@werdnanoslen are you still working on this? It seems that there are still bugs with the borders in the mobile variant. Both by default, on hover, and selected states.

**Review by jnm2377 — CHANGES_REQUESTED** [review]
I'm still seeing bugs: 
Screen Shot 2019-07-22 at 2 23 53 PM
Screen Shot 2019-07-22 at 2 23 42 PM
Screen Shot 2019-07-22 at 2 24 20 PM
Screen Shot 2019-07-22 at 2 24 12 PM

**Comment by werdnanoslen** [comment]
Yeah those border bugs and the hover effects I mentioned in my original post are things I still need help with @jnm2377. I haven't made any changes since then as I was waiting on this discussion to resolve.

**Comment by werdnanoslen** [comment]
Ok, I think I fixed a few bugs and also simplified the scss @jnm2377.

**Comment by tw15egan** [comment]
@werdnanoslen can you take a look at the merge conflicts?

**Comment by werdnanoslen** [comment]
@tw15egan resolved conflicts

**Comment by jnm2377** [comment]
Hi @werdnanoslen thanks so much for taking the time to get to know our codebase and contribute this PR. I am still seeing style bugs in the mobile example. Specifically, none of the borders are showing up on default, hover or select, and the expand caret needs to be aligned at the top not the middle. I understand that this is a very complicated component. We are interested in this variation but won't merge it with style bugs, and unfortunately feature enhancements are not on our roadmap right now, so we don't have dedicated developers on this. If you get the chance to work on this in the next week, we'd be happy to re-review. Otherwise, we'll close this in a week since it has been stale for a while. If you are able to work on this after it's been closed, we'd be more than happy to have you open another PR. We are just trying to avoid keeping stale PRs open for a long time.

**Comment by werdnanoslen** [comment]
@jnm2377 

> none of the borders are showing up on default, hover or select

Looks like some changes in the latest base branch merges messed with it. And the behavior I'm seeing is very very weird. 
  - Currently, The `td` border is set to 1px, but appears invisible, and is computed to be 0.5px (according to the dev console).
  - If I set it to 2px, then it becomes visible, but the `td` with expander carat magically gets an additional (and uneditable) border-top. And of course, now the border is not the same as other data table styles, which isn't cool.
  - If I set the borders to apply to the `td`s instead of `tr`s, then all the `td`s get a visible border but they are each different visual widths and computed (again, via dev console) to be 0.5px for expander `td`, 0px for checkbox `td` (again, still visible somehow??), and 1px for the content `td`.

I can make some shims, but honestly I don't know what to do here that is sustainable. Does anyone know what might be causing this behavior?

----

> the expand caret needs to be aligned at the top not the middle

I'm not sure what you mean here, I haven't changed the alignment at all from its expandable-table styling. It appears to me to be visually the same as the other examples:

image

image

**Pull request closed without merging**